### DISCUSSION TITLE
Why std::atomic cannot call many functions in IAsyncAction methods (all of them look like writing functions)?

### DISCUSSION THREAD

**Discussion by sxlllslgh** [microsoft/WindowsAppSDK#4273] [Q&A] [answered]
There are a vector storing my custom classes, which is declared in the main page class:
```cpp
class Foo {
    std::atomic<bool> flag;
};
```
```cpp
std::vector<Foo> myVector { Foo(), Foo(), ..., Foo() };
```
Then I want to execute a function for each element of `myVector` in parallel:
```cpp
for (auto& foo: myVector( {
    // I want to run this function in parallel, so I use an async function to return immediately.
    TimeCostFunctionOfThisClass(foo);
}

winrt::Windows::Foundation::IAsyncAction TimeCostFunctionOfThisClass(const Foo& foo) {
    if (foo.flag) {
        co_return;
    }
    // Compiler error, no "store" function of "foo.flag".
    foo.flag.store(true);
    // Some operations cost long time.
    foo.flag.store(false);
    co_return;
}
```
To be honest, I think my implementation is ugly. What I want is a process like this:
![diagram](https://github.com/microsoft/WindowsAppSDK/assets/11386258/ea3cf9f5-d427-4a51-8f3f-27b3523f2525)
In fact, I have tried OpenMP but it seems in each loop threads are re-created, so it took longer time. A more modification approach is to create threads when each `Foo` instance was created, but the overhead of thread synchronization and communication seems to be huge. If anyone can give some nifty suggestions I would be very grateful.

**Comment by DarranRowe** [comment, marked as answer]
Well, I would guess that the missing store is due to the fact that the function parameter is const. The const marks the entire Foo as constant, and the store modifies the flag member of Foo. This can't happen.
As long as you don't modify the vector until after all of the co routines return, then you don't need the const.

**Reply by sxlllslgh** [reply]
It seems true, I made a stupid mistake. Thank you very much!
